mirage/stub/dns_stub_mirage.ml

```diff
@@ -0,0 +1,401 @@
+      begin Client.get_resource_record t.client key name >|= function
+        | Error `Msg msg ->
+          Logs.err (fun m -> m "couldn't resolve %s" msg);
+          let data = `Rcode_error (Rcode.ServFail, Opcode.Query, None) in
+          inc `Resolver_servfail;
+          Some (build_reply data)
+        | Error `No_data (domain, soa) ->
+          let answer = (Name_rr_map.empty, Name_rr_map.singleton domain Soa soa) in
+          let data = `Answer answer in
+          inc `Resolver_nodata;
+          Some (build_reply data)
+        | Error `No_domain (domain, soa) ->
+          let answer = (Name_rr_map.empty, Name_rr_map.singleton domain Soa soa) in
+          let data = `Rcode_error (Rcode.NXDomain, Opcode.Query, Some answer) in
+          inc `Resolver_nodomain;
+          Some (build_reply data)
+        | Ok reply ->
+          let answer = (Name_rr_map.singleton name key reply, Name_rr_map.empty) in
+          let data = `Answer answer in
+          inc `Resolver_answers;
+          Some (build_reply data)
+      end
+    | _ ->
+      Logs.err (fun m -> m "not implemented %a, data %a"
+                   Dns.Packet.Question.pp question
+                   Dns.Packet.pp_data data);
+      let data = `Rcode_error (Rcode.NotImp, Packet.opcode_data data, None) in
+      inc `Resolver_notimp;
+      Lwt.return (Some (build_reply data))
+
+  (* we're now doing up to three lookups for each request:
+    - in authoritative server (Dns_trie)
+    - in reserved trie (Dns_trie)
+    - in resolver cache (Dns_cache)
+    - asking a remote resolver
+
+     instead, on startup authoritative (from external) could be merged with
+     reserved (but that makes data store very big and not easy to understand
+     (lots of files for the reserved zones)) *)
+  let handle t proto ip data =
+    match Packet.decode data with
+    | Error err ->
+      inc `Decoding_errors;
+      (* TODO send FormErr back *)
+      Logs.err (fun m -> m "couldn't decode %a" Packet.pp_err err);
+      Lwt.return None
+    | Ok packet ->
+      inc `Queries;
+      (* check header flags: recursion desired (and send recursion available) *)
+      let build_reply ?additional data =
+        let packet = Packet.create ?additional packet.header packet.question data in
+        fst (Packet.encode proto packet)
+      in
+      server t proto ip packet data build_reply >>= function
+      | Some data -> Lwt.return (Some data)
+      | None ->
+        (* next look in reserved trie! *)
+        let question, data = packet.Packet.question, packet.Packet.data in
+        match query_server t.reserved question data build_reply with
+        | Some data -> inc `Reserved_answers ; Lwt.return (Some data)
+        | None -> resolve t packet.Packet.question packet.Packet.data build_reply
+
+  let create ?(nameserver = Ipaddr.V4.of_string_exn "141.1.1.1") ?(size = 10000) ?(on_update = fun ~old:_ ?authenticated_key:_ ~update_source:_ _trie -> Lwt.return_unit) primary stack =
```

Comment:
what is this nameserver, could we document that in a comment and/or mli eventually?

Reply:
what motivates the `?size=10000` default, why not e.g. 65000 ?

Reply:
141.1.1.1 is an IPv4 address I remember -- an anycast recursive resolver run by cable&wireless. it should not be here (but iirc i encountered some issues connecting to censurfridns from marrakesh, so I put another nameserver in here) -- I agree with your above comment that we should have only a single IP hardcoded once :)

size was chosen arbitrarily... this is the number of resource records in the cache (while for use cases, I'd be more interested in being able to specify "number of bytes used by the cache").